From caa815070f1c0225f7ae718f9c6659bd8f2e7edf Mon Sep 17 00:00:00 2001
From: Zhang Qiao <zhangqiao22@huawei.com>
Date: Mon, 23 May 2022 17:47:44 +0000
Subject: [PATCH 2156/2592] [Huawei] sched/fair: Update rq clock before
 unthrottle a qos cfs_rq

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 186828

-------------------------------

------------[ cut here]------------
 rq->clock_update_flags < RQCF_ACT_SKIP
 WARNING: CPU: 5 PID: 3312 at kernel/sched/sched.h:1223 update_curr+0x1e5/0x210
 CPU: 5 PID: 3312 Comm: a.out Tainted: G S5.10.0.zq+ #1
 Hardware name: Huawei RH2288H V3/BC11HGSA0, BIOS 3.35 10/20/2016
 RIP: 0010:update_curr+0x1e5/0x210
 enqueue_entity+0x378/0xd00
 unthrottle_qos_cfs_rq+0x1bc/0x2a0
 __unthrottle_qos_cfs_rqs+0x87/0xa0
 qos_overload_timer_handler+0x35/0x60
 __run_hrtimer+0x5e/0x190
 __hrtimer_run_queues+0x81/0xe0
 hrtimer_interrupt+0x110/0x2c0
 __sysvec_apic_timer_interrupt+0x5f/0xd0
 sysvec_apic_timer_interrupt+0x31/0x80
 asm_sysvec_apic_timer_interrupt+0x12/0x20

After the last rq_pin_lock(), there is no rq clock update before calling
enqueue_entity() at unthrottle_qos_cfs_rq();

This patch fixes it by updating rq clock after holding rq->lock/pi_lock.

Fixes: c44b38d79334 ("[Huawei] sched/qos: Add qos_tg_{throttle,unthrottle}_{up,down}")
Signed-off-by: Zhang Qiao <zhangqiao22@huawei.com>
---
 kernel/sched/fair.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 072e591152ff..fb3f9e004b59 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -7398,6 +7398,7 @@ static void unthrottle_qos_cfs_rq(struct cfs_rq *cfs_rq)
 
 	cfs_rq->throttled = 0;
 
+	update_rq_clock(rq);
 	list_del_init(&cfs_rq->qos_throttled_list);
 
 	/* update hierarchical throttle state */
-- 
2.17.1

